### PR TITLE
Sort network services based on availability in the connect-to-vpn extension

### PR DESCRIPTION
**Pull request by rasmusbe**
## Description

Adds an preference to hide disabled services in the VPN list.

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Comment by pernielsentikaer** [comment]
Hi @rasmusbe 🔥 

Do you mind adding a changelog entry?

@sato11 do you mind checking this 🙂

**Comment by sato11** [comment]
Hi, thanks for the proposal and the heads-up. While I do not mind reviewing this, I am not yet able to imagine what practical situation has driven you to enable this feature. Perhaps you start to see a set of disabled services in your panel and they annoy you? It sounds legit but this leads to another question: in what scenario do we want to switch preferences so that disabled services pop up again?

To tell you the truth I had never seen a "disabled" network service when I wrote this extension, and not even after that. That's why they aren't filtered at all. But now it seems the case is different to you, and I'd like to hear a little more what an issue is.

If you reckon hiding the disabled is a feature which should default to true, I do not mind rendering it an opinionated feature with no interactions with preferences. Or if you are seeing a couple of disabled services at the very top of your item list and it's getting on your nerves that you need to press down arrow every time, sorting them so the disabled items go to the bottom could be also an option.

I do not mind if this gets merged as it is. But if you want me to review it'd be a lot more helpful if you share with us what issue you want to resolve by merging this patch.

**Comment by pernielsentikaer** [comment]
I never tried the application myself so I'm glad you have the right questions @sato11, do you mind enlighten us @rasmusbe 🙂

**Comment by rasmusbe** [comment]
Hi,

When I run the command I get some disabled services in the list. This PR adds an option to remove them.
image

I will fix the changelog later today 😄

**Comment by rasmusbe** [comment]
Changelog added ✅

File: extensions/connect-to-vpn/src/index.tsx [hunk 1]
```diff
@@ -105,6 +105,12 @@ export default function Command() {
 
         Promise.all(promises).then((services) => {
           const networkServices = services.reduce((acc, service) => {
+            if (getPreferenceValues().hideDisabled && service.status == "invalid") {
+              // Remove notice about disabled services
+              setHeader(undefined);
+              return acc;
+            }
+
```
Comment (sato11) [inline extensions/connect-to-vpn/src/index.tsx:113]: Thanks for the response at https://github.com/raycast/extensions/pull/7004#issuecomment-1588945552. If I were in your shoes I would not place the logic in the middle of reduce wrapped in promises as it'd make it harder to make sense. I would rather tweak only presentational logic in render. See the suggestion that I post aside.

**Comment by sato11** [comment]
This is the code I propose that it should be. Instead of preserving conditional branching in render() and making the data bypass it in reduce, I'd just change the way how data should be displayed without touching the underlying data store. 

It should appear that it has a lot of boilerplates, but I bet they are easier to extract or abstract from than the case where the underlying data take different form depending on service status.

I believe this is another way to resolve your issue. How does it sound to you? 🏓 

<details>

```diff
diff --git a/extensions/connect-to-vpn/src/index.tsx b/extensions/connect-to-vpn/src/index.tsx
index 476f5f97c..9e8f6f4aa 100644
--- a/extensions/connect-to-vpn/src/index.tsx
+++ b/extensions/connect-to-vpn/src/index.tsx
@@ -1,6 +1,6 @@
-import { List, showToast, Toast, ActionPanel, Action, Icon } from "@raycast/api";
-import { useEffect, useState } from "react";
+import { Action, ActionPanel, Icon, List, Toast, getPreferenceValues, showToast } from "@raycast/api";
 import { exec } from "child_process";
+import { useEffect, useState } from "react";
 
 export default function Command() {
   const [isLoading, setIsLoading] = useState(true);
@@ -127,37 +127,45 @@ export default function Command() {
   return (
     <List isLoading={isLoading}>
       <List.Section title={header}>
-        {Object.values(networkServices).map((service) => {
-          if (service.status == "connected") {
-            return (
-              <List.Item
-                icon={Icon.Checkmark}
-                key={service.id}
-                title={service.name}
-                actions={
-                  <ActionPanel>
-                    <Action title="Disconnect" onAction={() => disconnectFromPPPoEService(service)} />
-                  </ActionPanel>
-                }
-              />
-            );
-          } else if (service.status == "disconnected") {
-            return (
-              <List.Item
-                icon={Icon.Circle}
-                key={service.id}
-                title={service.name}
-                actions={
-                  <ActionPanel>
-                    <Action title="Connect" onAction={() => connectToPPPoEService(service)} />
-                  </ActionPanel>
-                }
-              />
-            );
-          } else {
-            return <List.Item icon={Icon.XmarkCircle} key={service.id} title={service.name} />;
-          }
-        })}
+        {/* connected services show up first */}
+        {Object.values(networkServices)
+          .filter(service => service.status == "connected")
+          .map((service) =>
+            <List.Item
+              icon={Icon.Checkmark}
+              key={service.id}
+              title={service.name}
+              actions={
+                <ActionPanel>
+                  <Action title="Disconnect" onAction={() => disconnectFromPPPoEService(service)} />
+                </ActionPanel>
+              }
+            />
+          )}
+        {/* then disconnected i.e. eligible services */}
+        {Object.values(networkServices)
+          .filter(service => service.status == "disconnected")
+          .map((service) =>
+            <List.Item
+              icon={Icon.Circle}
+              key={service.id}
+              title={service.name}
+              actions={
+                <ActionPanel>
+                  <Action title="Connect" onAction={() => connectToPPPoEService(service)} />
+                </ActionPanel>
+              }
+            />
+          )
+        }
+        {/* the rest is irrelevant and placed on the bottom */}
+        {/* I do not mind not displaying them at all depending on a preference, but since the underlying api responds I'd leave them as they are */}
+        {Object.values(networkServices)
+          .filter((service) => service.status == "invalid")
+          .map((service) => {
+            <List.Item icon={Icon.XmarkCircle} key={service.id} title={service.name} />;
+          })
+        }
       </List.Section>
     </List>
   );
@@ -175,4 +183,4 @@ type NetworkServices = {
   [id: string]: NetworkService;
 };
 
-type NetworkServiceStatus = "connected" | "disconnected";
+type NetworkServiceStatus = "connected" | "disconnected" | "invalid";
```

</details>

**Comment by rasmusbe** [comment]
@sato11 sound good, my only question is if only a sort function is needed in that case. So sort connected services first, then disconnected and disabled. That modifies the data a bit, but keeps everything available. What are your thoughts about that?

**Comment by sato11** [comment]
I do not think we need any explicit sort with the snippet I have provided above. See how it dispenses with sorting and (hopefully) satisfies your need. Did you try it in your development environment? If you have already tried it out and found a flaw in there, there might be a further chance for enhancement though. For example I'm not sure if it works alright when you have three available items in the list and you enable either the second or third item so they start to have connected status which should elevate them to the top of the list. Does it feel comfortable interacting with when reordering takes place? 

Generally speaking, I would consider modification of source data as the final resort, not the first choice, and I tend to devise the way out in the logic closer to users. However, it does not mean you are prohibited to choose to use any sorts at all. If it enhances the UX for example, we should perhaps be using it. 

So please do not hesitate to do what pleases you, but provide us with specific reasons why you want to do it and persuade us. I will send you feedbacks, if there is any, no matter what.

**Comment by rasmusbe** [comment]
@sato11 any feedback on this? 🏓

**Comment by pernielsentikaer** [comment]
Do you mind checking once more? @sato11 🔥

**Comment by sato11** [comment]
Hey, sorry for being very late. I've been thru some personal events since the end of June and not able to have energy enough to browse and respond to notifications. I'll be ready to resume next week, so please kindly give me a couple more days. Thanks.

**Review by sato11 — APPROVED** [review]
My style of work has undergone a series drastic changes since when I first reviewed these changes. VPNs having become non-requirement for me to do my job, I have got rid of them at all and now I find myself unable to test these changes on an actual environment. I believe your patch is awesome, fulfilling what you want the extension to do. But for the reasons I mentioned above, I guess I no longer qualify as the best reviewer. Hence I have left a comment regarding authorship of this extension. Otherwise, it looks good to me.

I thank you for your enduring patience, @rasmusbe. Now I'm ready to reply if you have comments.

**Comment by rasmusbe** [comment]
Thank you @sato11, I understand! 
I will continue to maintain the plugin in your absence.

**Comment by rasmusbe** [comment]
@pernielsentikaer do I need to do any changes to the CODEOWNERS file or does it update automatically?

**Comment by pernielsentikaer** [comment]
Thanks for all your hard work @sato11 🔥 I hope everything is okay? Feel free to DM me if you need anything

I just updated the backend and are now ready to merge it @rasmusbe

**Review by pernielsentikaer — APPROVED** [review]
Hi 👋 

Looks good to me, approved 🔥

**Pull request merged**